### PR TITLE
allow a read-only rbd image mounted by multiple pods

### LINKED ISSUE #27725: NoDiskConflict failure with RBD read-only volumes
Only a single pod per node is able to use the same Ceph RBD image in read-only mode. Further pods fail to be scheduled with `NoDiskConflict`.

I would expect multiple pods per node to be able to use the same RBD image in read-only mode, for example by mapping the image once per node and bind-mounting it into each pod. Is there any reason why this isn't possible?

### PR DESCRIPTION
**Pull request by rootfs**
allow pod to run read-only rbd volume 
fix #27725

### CODE REVIEW

**Comment by rootfs** [comment]
@davidopp PTAL thanks!

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 1]
```diff
@@ -108,7 +108,8 @@ func isVolumeConflict(volume api.Volume, pod *api.Pod) bool {
 		if volume.RBD != nil && existingVolume.RBD != nil {
 			mon, pool, image := volume.RBD.CephMonitors, volume.RBD.RBDPool, volume.RBD.RBDImage
 			emon, epool, eimage := existingVolume.RBD.CephMonitors, existingVolume.RBD.RBDPool, existingVolume.RBD.RBDImage
-			if haveSame(mon, emon) && pool == epool && image == eimage {
+			// same RBD image mounted by multiple Pods conflicts unless all Pods mount the image read-only
+			if haveSame(mon, emon) && pool == epool && image == eimage && !(volume.RBD.ReadOnly && existingVolume.RBD.ReadOnly) {
```
Comment (bprashanth) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: so the other checks don't matter for read only? if so, can we just check the rw mode on the RDB instead? or please add a comment about why the same mon/emon/pool won't work for read write and why it isn't an issue for read only (I'm pretty new to ceph).

**Comment by rootfs** [comment]
@bprashanth comments added about other checks. Thanks.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 2]
```diff
@@ -108,7 +108,10 @@ func isVolumeConflict(volume api.Volume, pod *api.Pod) bool {
 		if volume.RBD != nil && existingVolume.RBD != nil {
 			mon, pool, image := volume.RBD.CephMonitors, volume.RBD.RBDPool, volume.RBD.RBDImage
 			emon, epool, eimage := existingVolume.RBD.CephMonitors, existingVolume.RBD.RBDPool, existingVolume.RBD.RBDImage
-			if haveSame(mon, emon) && pool == epool && image == eimage {
+			// two RBDs images are the same if they share the same Ceph monitor, are in the same RADOS Pool, and have the same image name
+			// only one read-write mount is permitted for the same RBD image.
+			// same RBD image mounted by multiple Pods conflicts unless all Pods mount the image read-only
```
Comment (bprashanth) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: what about one mount as r/w and everything else as r/o? that feels useful for data pipelines. Just ack/nack and I'm ready to lgmt.

Comment (rootfs) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: It is possible to have one r/w and multiple r/o but it is tricky to get data sync on multiple mounts. IMO, we should solve this use case by introducing snapshot (to be implemented).

**Comment by Davidrjx** [comment]
in my case, Ceph RBD pvc with ReadWriteOnce can be accessed/mounted by multiple pods on same node, is that expected?

**Pull request merged**